python/triqs/lattice/utils.py

```diff
@@ -246,3 +246,161 @@ def TB_from_pythTB(ptb):
+    Hk: NumPy Array
+        the hamiltonian of the tight-binding model in momentum space in reduced analytical form
+
+    """
+
+    # imaginary number
+    I = sp.I
+
+    # matrix from the axis directions in momentum space
+    kx, ky, kz = sp.symbols("kx ky kz", real = True)
+    k_space_matrix = sp.Matrix([kx, ky, kz])
+
+    # symbolic dot product representation between the lattice unit vectors
+    # and the momentum space matrix
+    a1k, a2k, a3k = sp.symbols("a1k a2k a3k", real = True)
+    lattice = sp.Matrix([a1k, a2k, a3k])
+
+    # the number of orbitals involved in the electron hoppings
+    num_orb = tb_lat_obj.n_orbitals
+
+    # dictionary containing details about the hopping of the electrons
+    TB_lat_obj_hops = tb_lat_obj.hoppings 
+
+    # maximum hopping distances of electrons in each of the axial directions
+    max_x, max_y, max_z = list(np.max(np.array(list(TB_lat_obj_hops.keys())), axis = 0))
+
+    # number of cells involved in the hopping of electrons in each of the axial directions
+    num_cells_x, num_cells_y, num_cells_z = [2 * max_coord + 1 for max_coord in [max_x, max_y, max_z]]
+    
+    # basis of the 5D tensor real-space Hamiltonian
+    Hrij = np.zeros((num_cells_x, num_cells_y, num_cells_z, num_orb, num_orb), dtype = sp.exp)
+
+    # looping through the hopping parameters of the electrons involved in the inter-orbital hoppings
+    # key represents the cell coordinates of where the electrons hop to relative to the home unit cell
+    # hopping represents the matrix with the embedded hopping amplitudes
+    for key, hopping in TB_lat_obj_hops.items():
+        rx, ry, rz = key
+        # reduce floating point precision of hopping parameters to 3 decimal places
+        hopping = np.around(hopping, decimals = 3)
+        Hrij[rx + max_x, ry + max_y, rz + max_z] = hopping
+
+    # basis of the exponential term in the calculation of Hk
+    Hexp = np.empty_like(Hrij, dtype = sp.exp)
+
+    # perform the Fourier transform
+    for xi, yi, zi in itp(range(num_cells_x), range(num_cells_y), range(num_cells_z)):
+        coefficients = np.array([xi - max_x, yi - max_y, zi - max_z])
+        r = lattice.dot(coefficients)
+        eikr = sp.exp(-I * r)
+        Hexp[xi, yi, zi, :, :] = eikr
+
+    # summation over all real space axes
+    Hk = np.sum(Hrij * Hexp, axis = (0, 1, 2))
+    
+    # rewriting the exponential terms in the analytical expression in terms of 
+    for i, j in itp(range(num_orb), repeat = 2):
+        Hk[i, j] = Hk[i, j].rewrite(sp.cos)
+
+    # dealing with the numerical Hamiltonian
+    # we convert it to a SymPy matrix to use the substitutions method available in SymPy
+    Hk_numerical = sp.Matrix(Hk)
+    TB_lat_obj_units = tb_lat_obj.units
+    TB_lat_obj_units_transpose = np.transpose(TB_lat_obj_units)
+    
+    # obtaining unit vectors
+    # reduce floating point precision to 3 decimal places
+    a1 = np.around(TB_lat_obj_units_transpose[0], decimals = 3)
```

Comment:
does it make sense to let the user define the precision at this point?

Reply:
The default according to Wannier is 10^-6 but the user can change it, there will be an optional parameter to change the precision.